### PR TITLE
EDPDEV-896 Manage Shortcuts

### PR DESCRIPTION
**Pull request by nrasulicpfm**

### CODE REVIEW

**Review by davecap — CHANGES_REQUESTED** [review]
Looks good, a few minor changes to consider

File: solvebio/resource/object.py [hunk 1]
```diff
@@ -202,8 +202,9 @@ def get_or_create_by_full_path(cls, full_path, **kwargs):
         try:
             return cls.get_by_full_path(
                 full_path, assert_type=assert_type, client=_client)
-        except NotFoundError:
-            pass
+        except (NotFoundError, SolveError) as e:
+            if isinstance(e, NotFoundError) or "404" in e.message:
+                pass
```
[the remark this answers is not in the document]
Comment (nrasulicpfm) [reply solvebio/resource/object.py:207]: originally this was just 
```
except NotFoundError:
```

And that is working fine on vsim-dev. However when using qb-int-dev it raises a SolveError 404. That's why I added this to cover both not-found cases. But I'm not sure why this is different between the different environments - do you have an idea what could be the cause?

File: solvebio/resource/object.py [hunk 2]
```diff
@@ -382,6 +383,47 @@ def create_folder(cls, vault, full_path, tags=None, **kwargs):
...
+
+            target = {
+                'id': self.id,
+                'object_type': self.object_type
+            }
+
+            shortcut = Object.get_or_create_by_full_path(full_path=shortcut_full_path,
+                                                         object_type="shortcut",
+                                                         target=target)
```
[the remark this answers is not in the document]
Comment (nrasulicpfm) [reply solvebio/resource/object.py:421]: sorry, left it out when refactoring to use create_by_full_path. Fixed.

File: solvebio/resource/object.py [hunk 3]
```diff
@@ -382,6 +383,47 @@ def create_folder(cls, vault, full_path, tags=None, **kwargs):
...
+            if existing_object:
+                raise SolveError(
+                    "Object type {} already exists at location: {}".format(
+                        existing_object.object_type, full_path
+                    )
+                )
+        except (NotFoundError, SolveError) as e:
+            # Create the shortcut if an object doesn't exist on the path
+            # If any error other than 404 Not Found is caught - propagate it further
+            if isinstance(e, NotFoundError) or "404" in e.message:
```
[the remark this answers is not in the document]
Comment (nrasulicpfm) [reply solvebio/resource/object.py:411]: fixed

File: solvebio/resource/object.py [hunk 4]
```diff
@@ -594,6 +636,22 @@ def is_folder(self):
...
+        return self.object_type == 'shortcut'
+
+    def get_target(self):
+        if not self.is_shortcut:
+            raise SolveError("Only shortcut objects have a target resource. This is a {}".format(self.object_type))
+        target = self['target']
+        if target['object_type'] == 'url':
+            return target['url']
+        elif target['object_type'] == 'vault':
+            from . import Vault
```
[the remark this answers is not in the document]
Comment (nrasulicpfm) [reply solvebio/resource/object.py:650]: yep this works. Added additional handling for the special case that shortcut target is None - though this should never be the case in a regular scenario

File: solvebio/resource/object.py [hunk 5]
```diff
@@ -640,10 +636,16 @@ def is_file(self):
...
-    def get_target(self):
+    def get_target(self, return_none_target=True):
         if not self.is_shortcut:
             raise SolveError("Only shortcut objects have a target resource. This is a {}".format(self.object_type))
         target = self['target']
+        if not target:
+            if return_none_target:
+                return None
+            else:
+                raise SolveError("Shortcut target is None.")
```
Comment (davecap) [inline solvebio/resource/object.py:647]: Change to `Shortcut target not found`

Comment (davecap) [reply solvebio/resource/object.py:647]: Do we have a case where `return_none_target` would be False?

**Comment by davecap** [comment]
Overall looks good, just one more comment to address. Also need some docs to explain how to create and work with shortcuts. would be useful to also have a low level example that shows how to create a shortcut Object such as a URL.

**Review by davecap — APPROVED** [review]

**Comment by davecap** [comment]
Going to merge this as it seems to be feature complete.

**Pull request merged**